### ISSUE TITLE
Incorrect DrawContext.currentLayer usage

### ISSUE THREAD

**Issue by markpet49** [NASAWorldWind/WebWorldWind#494]
### Description

A number of classes incorrectly rely on dc.currentLayer during the ordered renderable drawing phase (Polygon, AbstractMesh, etc). They should instead rely on the layer property that is captured during the layer drawing phase. See the captured slack notes below:

### Additional Information

@markpet49:                
It would appear to me that AbstractMesh needs to account for the fact that it might not belong to a layer in the opacity calculations at around line 357 in AbstractMesh.doRenderOrdered. Am I correct in assuming that when a mesh is drawn as an ordered renderable that currentLayer is not valid? I believe this works by accident currently because WorldWindow.drawLayers leaves a dangling reference to currentLayer, so the mesh winds up using the opacity from the last layer drawn (usually ViewControlsLayer). In Miguels version, I clear currentLayer at the end of drawLayers which exposed the problem. Is my logic correct on this or is there some subtlety that I'm missing? Thanks.

@pdavidc:
Yes, you're correct, and thanks for cleaning up currentLayer at the end of drawLayers.
• DrawContext.currentLayer is defined during layer rendering, e.g. Layer.render and Renderable.render.
• DrawContext.currentLayer is undefined during ordered rendering. • Shapes keep track of the last layer they encountered during layer rendering.

**Comment by pdavidc** [comment]
@markpet49 An alternate method for fixing this is to ensure that dc.currentLayer is defined during ordered rendering. We could capture the current layer in addOrderedRenderable, then set it appropriately for each entry during ordered rendering. We do this now for globe state.

**Comment by markpet49** [comment]
When fixing this issue, please also address the TODO at the bottom of WorldWindow.drawLayers:
            // TODO: Following line commented out as workaround until issue #494 is addressed.
            // dc.currentLayer = null;

**Comment by pdavidc** [comment]
@zglueck This may be related to the recent work done to support layer opacity in shapes. Would you please look at usages of DrawContext.currentLayer and identify the places that are currently problematic?

**Comment by zglueck** [comment]
@markpet49  and @pdavidc I've reviewed uses of DrawContext.currentLayer and found some usage during ordered rendering operations, which exposes potential issues if the dangling layer reference does not correspond to the actual renderables layer.

All instances of invalid currentLayer use share a common pattern of actually assigning the current layer (during the render call) to a layer property attached to the ordered renderable. Some psuedo code to better illustrate:
```
render = function (dc) {
    var orderedRenderable = this.makeOrderedRenderable(dc);
    orderedRenderable.layer = dc.currentLayer;
};

makeOrderedRenderable = function (dc) {
     //configuration and setup
     return this;
};
```

So the simplest solution would be to replace the references to `dc.currentLayer` during ordered rendering with `this.layer`.

However, I feel this is a brittle design for two primary reasons.
1. Since the layer property is added in an ad-hoc manner in the render method, future development efforts may miss its availability (much as I did when refactoring opacity!).
2. The current pattern (demonstrated above) implicitly requires the return of the object (this) from `makeOrderedRenderable` which makes sense but feels a little fragile. While nothing has contradicted this approach in the library, nothing precludes further inheritance or development from missing the subtlety.

I'm leaning towards @pdavidc recommendation for defining currentLayer during ordered rendering. It alleviates renderabels from having to remember which layer is which and the implementation at first glance doesn't seem to be too complex. Further, as long as this ad-hoc layer property isn't utilized significantly, it could clean up the psuedo code pattern demonstrated above. I need to fully evaluate the use of the ad-hoc application of the layer property to be sure.

Barring any contradictive findings, I'll have a pr reflecting the proposed changes this evening/tomorrow morning.
